app/grandchallenge/core/utils/markdown.py

```diff
@@ -41,3 +43,29 @@ def run(self, root):
             if el.tag == "a":
                 el.set("target", "_blank")
                 el.set("rel", "noopener")
+
+
+class HtmlTagsExtension(Extension):
+    def extendMarkdown(self, md):  # noqa: N802
+        md.registerExtension(self)
+        md.postprocessors.register(
+            HtmlTagsPostprocessor(md), "htmltags_extension", 35
+        )
+
+
+class HtmlTagsPostprocessor(Postprocessor):
+    def run(self, lines):
+        for i in range(len(self.md.htmlStash.rawHtmlBlocks)):
+            bs4block = BeautifulSoup(
+                self.md.htmlStash.rawHtmlBlocks[i], "html.parser"
+            )
+            img = bs4block.find("img")
```

Comment:
Does this work for multiple images? Or does the processor receive one block at a time?

Reply:
It's going block by block it seems. Initialising Beautiful soup for every block in the passed html seems inefficient. Why do you use a `PostProcessor` and not a `TreeProcessor`? Why do you iterate over the blocks and not look at lines? It seems we now have two processors that do the same thing (HtmlTagsPostprocessor and BS4Treeprocessor), why is their functionality not combined?

Reply:
The HTML is extracted before processing the markdown and injected later after the processing, so it does not go throught the TreeProcessor. I thought of either using regex to get the img with classes, or to use BeautifoulSoup to parse the HTML. 
If I print the tags in the BS4Treeprocessor, I don't see the tags coming from the explicit HTML and that is why the class is not added to those tags in the first place.

Reply:
And  indeed, it is two processors now, one for the Markdown elements (through the TreeProcessor) and one for the raw HTML elements (through the PostProcessor). Any suggestions for an alternative?

Reply:
You can see [here](https://python-markdown.github.io/extensions/api/#usages)
![Screenshot 2024-08-14 at 11 53 37](https://github.com/user-attachments/assets/2aa9e2ba-5c61-4deb-aea0-4170c666841c)
So the raw HTML is removed before processing the Markdown

Reply:
My question is why do we have two bits of logic doing the same thing, you can see that you have duplicated functionality. Why is that logic not being reused?

Reply:
I see you point. I agree, the logic should be combined. However, as said earlier, the package does not pass the raw html through the same treeprocessor as the html from converted markdown. So, we still need two for loops to handle them, but now I made this processing in the same phase (TreeProcessor) to not worry about the execution order in the postprocessor phase, and combined the logic for setting the css class to remove the duplicated functionality. I made a commit.